src/Thorn_Calculus-Core_Bisimilarities.thy

```diff
@@ -1350,9 +1350,148 @@ text \<open>
   commutativity rules.
 *)
 
+lemma repeated_receive_evolution:
+  assumes "A \<triangleright>\<^sup>\<infinity> x. \<P> x \<rightarrow>\<^sub>s\<lparr>\<alpha>\<rparr> Q"
+  obtains n and X
+  where "\<alpha> = A \<triangleright> \<star>\<^bsup>n\<^esup> X"
+    and "Q = post_receive n X (\<lambda>v. \<P> v \<parallel> A \<triangleright>\<^sup>\<infinity> x. \<P> x)"
+  using assms
+  by
+    (subst (asm) (2) repeated_receive_proper_def)
+    (rotate_tac, induction \<alpha> "A \<triangleright> x. (\<P> x \<parallel> A \<triangleright>\<^sup>\<infinity> x. \<P> x)" Q rule: synchronous_transition.induct,
+     fastforce intro: repeated_receive_proper_def)
+
 lemma repeated_receive_idempotency [thorn_simps]:
   shows "A \<triangleright>\<^sup>\<infinity> x. \<P> x \<parallel> A \<triangleright>\<^sup>\<infinity> x. \<P> x \<sim>\<^sub>s A \<triangleright>\<^sup>\<infinity> x. \<P> x"
-  sorry
+proof (coinduction rule: synchronous.up_to_rule [where \<F> = "[\<sim>\<^sub>s] \<frown> \<M>"])
+  case (forward_simulation \<alpha> S)
+  then show ?case
+  proof cases
+    case (parallel_left_io \<eta> A' n X P')
+    from \<open>A \<triangleright>\<^sup>\<infinity> x. \<P> x \<rightarrow>\<^sub>s\<lparr>IO \<eta> A' n X\<rparr> P'\<close>
+    have "A \<triangleright> \<star>\<^bsup>n\<^esup> X = IO \<eta> A' n X" and "P' = post_receive n X (\<lambda>v. \<P> v \<parallel> A \<triangleright>\<^sup>\<infinity> x. \<P> x)"
```

Comment:
I think instead of proving `A ▹ ⋆⇗n⇖ X = IO η A' n X` it is better to prove what’s actually behind that, namely `η = Receiving` and `A = A'`.

Reply:
Also here, `v` should be replaced by `x`.

Reply:
Fixed in e284d0d6146420d94eba6d163bb939a47756ef4b and 47bd854e92b49c095c0e82208156126a1001e56a. In the latter, I extended the renaming of `v` to `x` to the whole proof of `repeated_receive_idempotency`. Please verify.

Reply:
Oh, I made a mistake above when writing `A = A'`. The statement should say that a variable introduced by `case` is equal to an already know one, not the other way round; so it should be `A' = A` instead. Could you please adapt the code correspondingly? Thanks.

Reply:
Somehow I’m still not really satisfied with this part of the code. I have the feeling it should be made more concise. Unfortunately, we don’t seem to have some other code that could serve as a blueprint, since this combination of case distinction and considering the only possible transition from a certain process family is kind of new.

Reply:
Regarding your last message, I'll think of a form to make that part more concise. With respect to the original issue in question, it's fixed in f5a7a4eaa290b44e3ad057c7d4eb0a4cda0d1415. By the way, I went ahead and also fixed the same issue for the `parallel_right_io` case. Please verify.

Reply:
Looks very good to me. Thanks for looking for a more concise version.

Reply:
Here’s a more concise one:
```isabelle
    case (parallel_left_io η A' n X P')
    from parallel_left_io(3,3)
    have "A ▹⇧∞ x. 𝒫 x →⇩s⦇A ▹ ⋆⇗n⇖ X⦈ post_receive n X (λx. 𝒫 x ∥ A ▹⇧∞ x. 𝒫 x)"
      by (fastforce elim: transition_from_repeated_receive)
```

Reply:
Also note how you can nicely pass a single premise from `case` twice: just repeat it’ index (3 in this case).

Reply:
I must say I pretty much like the second version.

Reply:
Sorry to steal the show from you. :slightly_frowning_face:

Reply:
Do you want to incorporate the second version into the code?

Reply:
Besides the nice trick to obtain `A ▹⇧∞ x. 𝒫 x →⇩s⦇A ▹ ⋆⇗n⇖ X⦈ post_receive n X (λx. 𝒫 x ∥ A ▹⇧∞ x. 𝒫 x)`, I still need to show that `"η = Receiving" and "A' = A" and "P' = post_receive n X (λx. 𝒫 x ∥ A ▹⇧∞ x. 𝒫 x)"`. And this only seems to be possible when using (again) `transition_from_repeated_receive` (which makes sense) and `parallel_left_io(3)` (again). Therefore, this is the new code I came up with:
```isabelle
    case (parallel_left_io η A' n X P')
    from parallel_left_io(3)
    have "η = Receiving" and "A' = A" and "P' = post_receive n X (λx. 𝒫 x ∥ A ▹⇧∞ x. 𝒫 x)"
      by (auto elim: transition_from_repeated_receive)
    with parallel_left_io(3)
    have "A ▹⇧∞ x. 𝒫 x →⇩s⦇A ▹ ⋆⇗n⇖ X⦈ post_receive n X (λx. 𝒫 x ∥ A ▹⇧∞ x. 𝒫 x)"
      by simp
```
 
Would you be satisfied with it?

Reply:
Okay, that’s not possible, because we need the equations as facts, as `η`, `A'`, and `P'` also appear in the goal. This seems to be a pretty good choice:
```isabelle
case (parallel_left_io η A' n X P')
from ‹A ▹⇧∞ x. 𝒫 x →⇩s⦇IO η A' n X⦈ P'›
have
  "η = Receiving"
and
  "A' = A"
and
  "P' = post_receive n X (λx. 𝒫 x ∥ A ▹⇧∞ x. 𝒫 x)"
and
  "A ▹⇧∞ x. 𝒫 x →⇩s⦇A ▹ ⋆⇗n⇖ X⦈ post_receive n X (λx. 𝒫 x ∥ A ▹⇧∞ x. 𝒫 x)"
  unfolding atomize_conj
  by (-, fastforce elim: transition_from_repeated_receive)
```

Reply:
That's a nice alternative indeed. I modified the code accordingly in 384b3e4cfef5b6e988e7579c4c484388391c857c. Also, I (hopefully) improved the case `receiving` in a similar manner.

Reply:
The change for the `receiving` case is a nice and important one indeed. It can be further improved by replacing `by -` with `.`. :slightly_smiling_face:

Reply:
That's right, I thought I had already tried with `.` and `..` and that they hadn't worked. The improvement is included in 0245d8188654cd3aaf4e4015e074ca721fd8a2e9, please verify.

Reply:
Looks good. :slightly_smiling_face: